File: binary_search/입국심사.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <queue>
#include <cmath>
#include <iostream>

using namespace std;


long long solution(int n, vector<int> times) {
    long long answer = 0;
    
    sort(times.begin(), times.end());
    
    long long min = 1;
    long long max = (long long)times[times.size()-1] * n;
    
    while(min <= max){
        long long mid = (min+max) / 2;
        long long total = mid * n;
        
        long long cur = 0;
        
        for(int i=0; i<times.size(); i++){
            cur += mid / (long long)times[i];
        }
        
        if(cur >= n){
            max = mid - 1;
            answer = mid;
        }
        else{
            min = mid + 1;
        }
    }
    
    return answer;
}

File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;

class Solution {
    public long solution(int n, int[] times) {
        
        Arrays.sort(times);
        long answer = Long.MAX_VALUE / 100;
        
        long low = 0;
        long high = Long.MAX_VALUE / 100;
        
        while(low < high){
            long mid = (low+high)/2;
            
            long cnt = 0;
            for(int i=0; i<times.length; i++){
                cnt += mid / times[i];
            }
            
            if(cnt >= n){
                high = mid;
                answer = mid < answer ? mid : answer;
            }
            else{
                low = mid+1;
            }
        }
        
        return low;
    }
}